rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java

```diff
@@ -140,26 +131,33 @@ public MavenPomDownloader(Map<Path, Pom> projectPoms, HttpSender httpSender, Exe
         this.projectPomsByGav = projectPomsByGav(projectPoms);
         this.httpSender = httpSender;
         this.ctx = MavenExecutionContextView.view(ctx);
-        this.sendRequest = Retry.decorateCheckedFunction(
-                mavenDownloaderRetry,
-                request -> {
-                    int responseCode;
-                    long start = System.nanoTime();
-                    try (HttpSender.Response response = httpSender.send(request)) {
-                        if (response.isSuccessful()) {
-                            return response.getBodyAsBytes();
-                        }
-                        responseCode = response.getCode();
-                    } catch (Throwable t) {
-                        throw new HttpSenderResponseException(t, null);
-                    } finally {
-                        this.ctx.recordResolutionTime(Duration.ofNanos(System.nanoTime() - start));
-                    }
-                    throw new HttpSenderResponseException(null, responseCode);
-                });
         this.mavenCache = this.ctx.getPomCache();
     }
 
+    byte[] sendRequest(HttpSender.Request request) throws HttpSenderResponseException {
+        long start = System.nanoTime();
+        try {
+            return Failsafe.with(retryPolicy).get(() -> {
+                try (HttpSender.Response response = httpSender.send(request)) {
+                    if (!response.isSuccessful()) {
+                        throw new HttpSenderResponseException(null, response.getCode());
+                    }
+                    return response.getBodyAsBytes();
+                }
+            });
+        } catch (Throwable t) {
```

Comment:
I wonder if it makes sense to catch `FailsafeException` separately, and if caught throw the underlying cause instead of wrapping that in a `HttpSenderResponseException`. I'll see if I can whip up an alternative.

Reply:
great minds and all. See latest commit